### ISSUE TITLE
CPU-side Segfault for FA4

### ISSUE THREAD

**Issue by howardzhang-cv** [Dao-AILab/flash-attention#2281]
# FA4 Intermittent Segfault in `cute.compile()` JIT Launcher on SM100 (GB200)

## Disclaimer
I was not able to reproduce the segfault just running a simpler script. Something about calling it specifically through this call stack is what causes it. It uses [this](https://github.com/pytorch/pytorch/pull/175472) particular PR (that is still a draft and is not part of PyTorch main branch yet). It only happens when I run it in the context of the Flux.1-schnell model.

## Summary

FA4 flash attention intermittently segfaults (~10-25% of the time) on NVIDIA GB200 (SM100, compute capability 10.0) during the first kernel launch after `cute.compile()`. The crash is **CPU-side** — `compute-sanitizer` finds zero GPU memory errors. GDB shows the crash at an unmapped address in JIT-compiled code (`0x0000ffff70001410 in ?? ()`).

The crashing kernel configuration is:
- **q/k/v**: `[1, 4608, 24, 128]` (BSHD layout), `torch.bfloat16`
- **causal**: `False`
- **q_stage**: `2` (because `seqlen_q=4608 > m_block_size=128`)
- **is_persistent**: `True` (non-causal, no varlen, no split-kv)
- **Kernel class**: `FlashAttentionForwardSm100`

### Minimal Repro Script

**`repro_fa4_segfault.py`** — loads FLUX.1-schnell, registers FA4, runs one inference pass:

```python
import faulthandler
faulthandler.enable()

import lpips
import torch
import torch._dynamo
from diffusers import FluxPipeline
from torch.nn.attention import (
    activate_flash_attention_impl,
    restore_flash_attention_impl,
)
from torchao.prototype.attention import (
    AttentionBackend,
    LowPrecisionAttentionConfig,
    apply_low_precision_attention,
)

@torch.inference_mode()
def main():
    print(f"GPU: {torch.cuda.get_device_name()}")
    print(f"Compute capability: {torch.cuda.get_device_capability()}")
    print(f"PyTorch: {torch.__version__}")
    print(f"CUDA: {torch.version.cuda}")

    pipe = FluxPipeline.from_pretrained(
        "black-forest-labs/FLUX.1-schnell", torch_dtype=torch.bfloat16
    )
    pipe = pipe.to("cuda")
    pipe.set_progress_bar_config(disable=True)

    # LPIPS VGG on GPU increases crash rate via memory pressure
    loss_fn = lpips.LPIPS(net="vgg").to("cuda")

    print(f"GPU memory: {torch.cuda.memory_allocated() / 1e9:.1f} GB")

    activate_flash_attention_impl("FA4")
    try:
        image = pipe(
            prompt="A red car",
            num_inference_steps=4,
            guidance_scale=3.5,
            generator=torch.Generator(device="cuda").manual_seed(42),
        ).images[0]
        torch.cuda.synchronize()
        print(f"OK: {image.size}")
    finally:
        restore_flash_attention_impl()

if __name__ == "__main__":
    main()
```

### Loop Script

Since the crash is intermittent (~10-25% per run), loop in fresh processes:

```bash
#!/bin/bash
MAX_RUNS="${1:-50}"
for i in $(seq 1 "$MAX_RUNS"); do
    echo "=== Run $i / $MAX_RUNS ==="
    python repro_fa4_segfault.py 2>&1
    EXIT_CODE=$?
    if [ "$EXIT_CODE" -ne 0 ]; then
        echo "CRASHED on run $i (exit code $EXIT_CODE)"
        exit 1
    fi
done
echo "No crash after $MAX_RUNS runs."
```

Typically crashes within the first 5-10 runs.

### Required Dependencies

```
torch                  (built from source, see commit below)
flash-attn-cute        (built from source, see commit below)
torchao                (built from source, see commit below)
diffusers              0.37.0.dev0
lpips                  0.1.4
nvidia-cutlass-dsl     4.4.0
cuda-bindings          13.1.1
triton                 3.6.0
```

FLUX.1-schnell model weights from `black-forest-labs/FLUX.1-schnell` (HuggingFace).

## Environment

| Component | Details |
|---|---|
| **GPU** | 2× NVIDIA GB200 (Blackwell, SM100, compute capability 10.0, 189 GB VRAM each) |
| **CPU** | ARM Neoverse-V2 (aarch64), 144 cores |
| **OS** | CentOS Stream 9 (kernel 6.11.1) |
| **NVIDIA Driver** | 570.133.20 |
| **CUDA Toolkit** | 12.9 (V12.9.86) |
| **Python** | 3.12.0 |
| **PyTorch** | 2.12.0a0+gitb3d2ce4 (source build from local checkout) |
| **PyTorch HEAD** | `437afb03e10` — includes "Add FA4 fp8 implementation to SDPA" patch |
| **flash-attention** | v2.8.3+279 commits, branch `fa4_fp8_support`, HEAD `a8c9512` |
| **nvidia-cutlass-dsl** | 4.4.0 |
| **torchao** | 0.17.0+git4a42d3227, branch `feature/low-precision-attn`, HEAD `1bca8f872` |
| **cuDNN** | Not detected (`None`) |
| **Conda env** | `oss_dev_flashv4` |

### Crash Rate vs. Process State

The crash probability depends on the process's memory layout. Various imports/loads shift allocation patterns and increase the rate:

| Process state | Crash rate | Evidence |
|---|---|---|
| FLUX model + all imports (torch._dynamo, lpips, torchao) | **~25%** | Crashed on run 1, 3, 4 |
| FLUX model + torch._dynamo + lpips + torchao | **~33%** | Crashed on run 3 |
| FLUX model + torch._dynamo only | **~1-2%** | Crashed on run 76 |
| FLUX model alone (no extra imports) | **<1%** | 0 crashes in 100 runs |
| Synthetic tensors only (no FLUX model) | **0%** | 0 crashes in 100 runs |

**None of these imports cause the crash** — they change the process memory layout to make the underlying bug more or less likely to trigger.

### Not FP8-Specific

The crash occurs with both **BF16** and **FP8** inputs. It was initially observed during FP8 attention benchmarking, but subsequent investigation confirmed BF16-only paths crash with the same symptoms.

### Not `is_persistent`-Specific

Setting `is_persistent=False` unconditionally in `FlashAttentionForwardSm100` did not prevent the crash.

### compute-sanitizer

```bash
compute-sanitizer --tool memcheck --print-limit 20 python repro_fa4_segfault.py
```

**Result**: 0 GPU memory errors detected. Process still segfaulted (exit code 11). This confirms the crash is **CPU-side**, not a GPU kernel memory violation.

### GDB Backtrace

```bash
gdb -batch -ex "set pagination off" -ex run -ex "thread apply all bt" -ex quit \
    --args python repro_fa4_segfault.py
```

**Result**:
```
Thread 1 received signal SIGSEGV, Segmentation fault.
0x0000ffff70001410 in ?? ()
#0  0x0000ffff70001410 in ?? ()
Backtrace stopped: Cannot access memory at address 0x0000ffffc93f0000
```

GDB cannot resolve the crash address because it's in **JIT-compiled code** from `cute.compile()` (no debug symbols). The "Cannot access memory" message indicates stack corruption or a jump to an invalid address in the JIT launcher.

### Debug Instrumentation

With `TORCHAO_FA4_DEBUG=1`, the crash occurs **after** `cute.compile() DONE` but **during** the first `compile_cache[key](*call_args)` invocation:

```
[FA4_DEBUG] cute.compile() START for new compile_key: is_fp8=False causal=False head_dim=128 seqlen_q=4608 q_stage=2
[FA4_DEBUG] cute.compile() DONE
[FA4_DEBUG] kernel FIRST RUN: q=[1, 4608, 24, 128] torch.bfloat16 k=[1, 4608, 24, 128] v=[1, 4608, 24, 128] out=[1, 4608, 24, 128] torch.bfloat16 is_fp8=False causal=False local=False pack_gqa=False is_split_kv=False q_stage=2 m_block=128 n_block=128 num_splits=1 seqlen_q=4608 seqlen_k=4608 num_head=24 num_head_kv=24 head_dim=128 head_dim_v=128 compute_capability=10
!!!!!!! Segfault encountered !!!!!!!
```

The `cute.compile()` call succeeds. The crash is in the compiled launcher function's first invocation.

## Call Chain

The crash occurs in the FA4 forward path:

```
F.scaled_dot_product_attention(q, k, v, is_causal=False)
  → aten::_scaled_dot_product_flash_attention [CUDA dispatch override]
  → _fa4_scaled_dot_product_flash_attention_forward_impl()  [torch/nn/attention/_fa4.py]
    → _transpose_dense(q, k, v)  # BHSD → BSHD (non-contiguous views)
    → out_bhsd = torch.empty_like(query)  # BHSD contiguous
    → out_bshd = out_bhsd.transpose(1, 2)  # BSHD non-contiguous view
    → _fa4_run_forward(q, k, v, out=out_bshd)
      → _flash_attn_fwd(q, k, v, out=out_bshd, causal=False)  [flash_attn/cute/interface.py]
        → FlashAttentionForwardSm100(head_dim=128, q_stage=2, is_persistent=True, ...)
        → cute.compile(fa_fwd, ...)  # JIT compile — succeeds
        → compile_cache[key](*call_args)  # ← CRASH HERE (first invocation)
```

## Crashing Kernel Parameters

From `FlashAttentionForwardSm100` constructor:

| Parameter | Value |
|---|---|
| `head_dim` | 128 |
| `head_dim_v` | 128 |
| `qhead_per_kvhead` | 1 (MHA, not GQA) |
| `is_causal` | False |
| `is_local` | False |
| `is_split_kv` | False |
| `pack_gqa` | False |
| `m_block_size` | 128 |
| `n_block_size` | 128 |
| `q_stage` | 2 |
| `is_persistent` | True |
| `is_varlen_q` | False |
| `paged_kv_non_tma` | False |
| `score_mod` | None |
| `mask_mod` | None |

## What We Ruled Out

| Hypothesis | How tested | Result |
|---|---|---|
| GPU kernel memory error | `compute-sanitizer --tool memcheck` | 0 errors found |
| FP8-specific bug | Crash observed with BF16-only path | BF16 crashes too |
| `is_persistent=True` tile scheduler | Forced `is_persistent=False` | Still crashes |
| Triton quantization kernels | BF16 path doesn't use Triton | Crashes without Triton |
| `cute.compile()` itself | Debug prints show "compile DONE" before crash | Compilation succeeds |
| Contiguous vs. non-contiguous tensors | Direct call with contiguous BSHD tensors | No crash in 100 runs (but non-contiguity is not the cause — the crash is memory-layout-sensitive) |
| Specific torch.compile/Inductor context | Crash occurs without torch.compile | Not required |

## Hypothesis

The crash is in the **JIT-compiled kernel launcher** generated by `cute.compile()` (nvidia-cutlass-dsl 4.4.0) for the `FlashAttentionForwardSm100` kernel on SM100/aarch64. The intermittency suggests the launcher has a latent memory corruption bug (e.g., buffer overflow, misaligned access, or use of an uninitialized pointer) that only manifests when specific memory addresses are assigned by the CUDA allocator or system allocator — which varies based on process memory state.

The fact that loading more models/libraries into the process (torch._dynamo hooks, LPIPS VGG weights, torchao custom op registration) significantly increases the crash rate supports this: these change the base addresses of subsequent allocations, making the latent bug more or less likely to trigger.

**Comment by tridao** [comment]
Thanks for the bug report. Can you try with cutlass-dsl 4.4.0.dev1 to see if it still crashes. There was some changes from 4.4.0.dev1 to 4.4.0

**Comment by howardzhang-cv** [comment]
> Thanks for the bug report. Can you try with cutlass-dsl 4.4.0.dev1 to see if it still crashes. There was some changes from 4.4.0.dev1 to 4.4.0

Thanks! It worked!
